src/pretty.rs

```diff
@@ -0,0 +1,115 @@
+                            ("SUCC", 1) => {
+                                increments += 1;
+                                arg = &args[0]
+                            }
+                            ("ZERO", 0) | ("0", 0) => return increments.to_string(),
+                            // number does not terminate with ZERO, so we use the
+                            // non-special-case printing style
+                            _ => break,
+                        }
+                    }
+                }
+                (".", 2) => {
+                    let mut first = &args[0];
+                    let mut rest = vec![&args[1]]; // in reverse order for fast `push`ing
+                    while let Some((op, args)) = first.as_application() {
+                        match (op.display(sig).as_str(), args.len()) {
+                            (".", 2) => {
+                                first = &args[0];
+                                rest.push(&args[1]);
+                            }
+                            _ => break,
+                        }
+                    }
+                    rest.push(first);
+                    rest.reverse();
+                    let interior = rest.into_iter()
+                        .map(|x| x.pretty_inner(sig, false))
+                        .join(" ");
+                    return if spaces_allowed {
+                        interior
+                    } else {
+                        format!("({})", interior)
+                    };
+                }
+                ("CONS", 2) => {
+                    let mut items = vec![&args[0]];
+                    let mut cdr = &args[1];
+                    while let Some((op, args)) = cdr.as_application() {
+                        match (op.display(sig).as_str(), args.len()) {
+                            ("CONS", 2) => {
+                                items.push(&args[0]);
+                                cdr = &args[1];
+                            }
+                            ("NIL", 0) | ("NULL", 0) => {
+                                return format!(
+                                    "[{}]",
+                                    items
+                                        .into_iter()
+                                        .map(|item| item.pretty_inner(sig, true))
+                                        .join(", ")
+                                )
+                            }
+                            // list does not terminate with NIL, so we use the
+                            // non-special-case printing style
+                            _ => break,
+                        }
+                    }
+                }
+                _ => (),
+            }
+            let args_str = args.iter()
+                .map(|arg| arg.pretty_inner(sig, true))
+                .join(", ");
+            format!("{}({})", op_str, args_str)
+        } else {
+            self.display(sig)
+        }
+    }
```

Comment:
`pretty_inner` might be easier to read and reason about if the special cases were factored out.

Reply:
See the latest commit. I'm not entirely sure whether I prefer it with or without the special cases pulled out — it might actually be easier to read with them altogether. I'll defer that decision to you